### ISSUE TITLE
LNK2001 error on compiling with visual studio 2017

### ISSUE THREAD

**Issue by Loller06** [yourWaifu/sleepy-discord#94]
Today i wanted to try this library and after several hours of compiling, copying, downloading, etc. i was able to set up the environment. So i wanted to test the library by copying and pasting the test code from the readme into a cpp file and, after all the linking directories, libraries and input thingies, i was ready to build. When i tried to build, those errors showed up (sorry the logs are italian because i am italian):
```
1>------ Inizio compilazione: Progetto: Discord Bot, Configurazione: Release Win32 ------
1>Origine.obj : error LNK2001: simbolo esterno "private: virtual void __thiscall SleepyDiscord::WebsocketppDiscordClient::runAsync(void)" (?runAsync@WebsocketppDiscordClient@SleepyDiscord@@EAEXXZ) non risolto
1>Origine.obj : error LNK2001: simbolo esterno "private: virtual void __thiscall SleepyDiscord::WebsocketppDiscordClient::send(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >)" (?send@WebsocketppDiscordClient@SleepyDiscord@@EAEXV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z) non risolto
1>Origine.obj : error LNK2001: simbolo esterno "private: virtual void __thiscall SleepyDiscord::WebsocketppDiscordClient::disconnect(unsigned int,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >)" (?disconnect@WebsocketppDiscordClient@SleepyDiscord@@EAEXIV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z) non risolto
1>Origine.obj : error LNK2001: simbolo esterno "private: virtual bool __thiscall SleepyDiscord::WebsocketppDiscordClient::connect(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (?connect@WebsocketppDiscordClient@SleepyDiscord@@EAE_NABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z) non risolto
1>Origine.obj : error LNK2001: simbolo esterno "protected: virtual void __thiscall SleepyDiscord::WebsocketppDiscordClient::sleep(unsigned int)" (?sleep@WebsocketppDiscordClient@SleepyDiscord@@MAEXI@Z) non risolto
1>Origine.obj : error LNK2001: simbolo esterno "protected: virtual void __thiscall SleepyDiscord::WebsocketppDiscordClient::onError(enum SleepyDiscord::ErrorCode,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >)" (?onError@WebsocketppDiscordClient@SleepyDiscord@@MAEXW4ErrorCode@2@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z) non risolto
1>Origine.obj : error LNK2001: simbolo esterno "public: virtual struct SleepyDiscord::Timer __thiscall SleepyDiscord::WebsocketppDiscordClient::schedule(class std::function<void __cdecl(void)>,__int64)" (?schedule@WebsocketppDiscordClient@SleepyDiscord@@UAE?AUTimer@2@V?$function@$$A6AXXZ@std@@_J@Z) non risolto
1>Origine.obj : error LNK2001: simbolo esterno "public: virtual void __thiscall SleepyDiscord::WebsocketppDiscordClient::run(void)" (?run@WebsocketppDiscordClient@SleepyDiscord@@UAEXXZ) non risolto
1>Origine.obj : error LNK2001: simbolo esterno "public: __thiscall SleepyDiscord::WebsocketppDiscordClient::~WebsocketppDiscordClient(void)" (??1WebsocketppDiscordClient@SleepyDiscord@@QAE@XZ) non risolto
1>Origine.obj : error LNK2001: simbolo esterno "public: __thiscall SleepyDiscord::WebsocketppDiscordClient::WebsocketppDiscordClient(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,char)" (??0WebsocketppDiscordClient@SleepyDiscord@@QAE@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@D@Z) non risolto
1>sleepy_discord.lib(client.obj) : error LNK2001: simbolo esterno "public: static class SleepyDiscord::GenericSession * (__cdecl* SleepyDiscord::CustomSession::init)(void)" (?init@CustomSession@SleepyDiscord@@2Q6APAVGenericSession@2@XZA) non risolto
1>C:\Users\dnbia\Documents\Cpp projects\Discord Bot\Release\Discord Bot.exe : fatal error LNK1120: 11 esterni non risolti
1>Compilazione progetto "Discord Bot.vcxproj" NON COMPLETATA.
========== Compilazione: 0 completate, 1 non riuscite, 0 aggiornate, 0 ignorate ==========
```
I tried everything! Adding libraries, directories, tried to recompile some things and so on but no luck! Please help!

**Comment by yourWaifu** [comment]
It looks like you are missing websocketpp

**Comment by Loller06** [comment]
i'm sure that websocketpp was downloaded correctly and the folder is there

**Comment by yourWaifu** [comment]
Can you send your list of include directories r for your bot?

**Comment by Loller06** [comment]
Here you go:
C:\Users\<myuser>\Documents\Cpp projects\sleepy-discord
C:\Users\<myuser>\Documents\Cpp projects\sleepy-discord\deps
C:\Users\<myuser>\Documents\Cpp projects\sleepy-discord\deps\include
C:\Users\<myuser>\Documents\Cpp projects\sleepy-discord\deps\lib
C:\Users\<myuser>\Documents\Cpp projects\sleepy-discord\include\sleepy_discord
C:\Users\<myuser>\Documents\Cpp projects\sleepy-discord\include\sleepy_discord\IncludeNonexistent

**Comment by yourWaifu** [comment]
If you didn't make any changes to the project properties. Then it's most likely not an issue with your project properties but most likely issues with you deps folder.

**Comment by Loller06** [comment]
Yeah I cleaned and tried to recompile sleepy discord and now I'm getting more than 100 errors, idk why. Can't you just give some precomputed builds with all the deps in it?

**Comment by yourWaifu** [comment]
I have directions for contacting me on Discord here https://yourwaifu.github.io/sleepy-discord/documentation.html#a-warm-welcome . We need to be on a common server to direct message each other on Discord so that's what Discord API server is for.

**Comment by yourWaifu** [comment]
No I can't, because of the many different setups and configurations that sleepy discord can be in. Maybe in the future, I can as most people don't use custom configurations. However maybe in the future once I can figure out how to do it automatically.

**Comment by SalmonSeasoning** [comment]
I am having the same issue on a slight smaller scale. (62 Errors + 1 Warning).
Repository: [https://github.com/SalmonSeasoning/DiscordRPCTest](https://github.com/SalmonSeasoning/DiscordRPCTest)
For a more simplistic view of my VS library configurations, take a look at [LIBRARIES](https://github.com/SalmonSeasoning/DiscordRPCTest/blob/master/LIBRARIES).
It complains about 60+ unresolved variables all apparently originating from your library.
Everything is configured for building on either Win32 ("x86") or Win64 ("x64").
Nothing has been modified so the exact result from my build should appear on your build as well.

**Comment by SalmonSeasoning** [comment]
I have done as you said. I trashed the old library, I cloned the repo from master branch, ran setup.py again, and integrated the IncludeNonexistant folder into the project with all available headers. However, all of this did not resolve the 62 errors that pop up when including the library itself. I then tried to import "default_functions.cpp" and it came down to just 2 errors that did not resolve.
```
1>------ Build started: Project: DiscordWindowsBot, Configuration: Debug Win32 ------
1>Source.cpp
1>d:\workspaces\discordwindowsbot\externals\sleepy_discord\sleepy_discord\default_functions.cpp(1): fatal error C1083: Cannot open include file: 'client.h': No such file or directory
1>Done building project "DiscordRPC.vcxproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```
I think the issue is that the file cannot find the client header when compiled and/or it's not being included when compiled at all.

Thanks for your reply,

also it's a throw-away application. I went ahead and changed the token anyway.

**Comment by yourWaifu** [comment]
@SalmonSeasoning I tried to find sleepy_discord.lib but couldn't find it but I ran the setup.py script for both websocketpp and cpr and compiled curl. I then compiled sleepy_discord.lib. I build your bot on the release configuration because your debug configuration wasn't even linking my library. I then ran into issues with your code and fixed them by removing them. I also ran into issues with the compiler not outputing obj files, and somehow fixed it by playing around with the project properties. I then added IncludeNonexistent to the include paths. I also added libcurl_a.lib, Ws2_32.lib, crypt32.lib, wldap32.lib, and Normaliz.lib. And it worked, didn't test your bot tho.

**Comment by SalmonSeasoning** [comment]
I've done close to exactly as you said (websocketpp, cpr, curl installed). I had to leave out libcurl_a.lib because the universal version of the library I have doesn't even have a compiled .lib at all. However, at this point, I'm sure it's my system being buggy so I'm just going to scratch and restart the repository with all of the same code and see if that works (especially since a GLFW3 .lib error popped out of nowhere). I think OP's issue is that he hasn't gone into the Sleepy Discord .sln and built for x64 (Debug) and included the .lib into his project solution.
(I tried it -both x86 and x64- and actually had an error pop up saying it was built for x86 and not x64 and building it for the "Release" just threw a bunch of errors regardless.)
